AK/InsertionSort.h

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2022, Marc Luqué <marc.luque@outlook.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Concepts.h>
+#include <AK/StdLibExtras.h>
+
+namespace AK {
+
+// Standard Insertion Sort, with `end` inclusive!
+template<typename Collection, typename Comparator, typename T = decltype(declval<Collection>()[declval<int>()])>
+void insertion_sort(Collection& col, int start, int end, Comparator comparator) requires(Indexable<Collection, T>)
+{
+    for (int i = start; i <= end; ++i) {
+        for (int j = i; j > 0 && comparator(col[j], col[j - 1]); --j)
+            swap(col[j], col[j - 1]);
+    }
+}
+
+// `end` is exclusive, as usual with iterators.
+template<typename Iterator, typename Comparator, typename T = decltype(*declval<Iterator>())>
+void insertion_sort(Iterator start, Iterator end, Comparator comparator) requires(IteratorType<Iterator, T>)
```

Comment:
Technically this isn't correct, the begin and end iterators can be of different types, see IteratorPairWih

Reply:
It does compile though :^)

Reply:
for the specific things you've tested it on, yes.
Try sorting a hashtable for instance (even though that makes no sense)